File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/NonDeterministicProcessesSpec.scala
package coop.rchain.rholang.interpreter.accounting

import cats.Parallel
import cats.data.Chain
import cats.effect._
import cats.mtl.FunctorTell
import cats.syntax.all._
import coop.rchain.crypto.hash.Blake2b512Random
import coop.rchain.metrics
import coop.rchain.metrics.{Metrics, NoopSpan, Span}
import coop.rchain.models.{BindPattern, ListParWithRandom, Par, TaggedContinuation}
import coop.rchain.rholang.Resources
import coop.rchain.rholang.interpreter.RhoRuntime.RhoHistoryRepository
import coop.rchain.rholang.interpreter.SystemProcesses.Definition
import coop.rchain.rholang.interpreter.accounting.utils._
import coop.rchain.rholang.interpreter.{EvaluateResult, _}
import coop.rchain.rholang.syntax._
import coop.rchain.rspace.RSpace.RSpaceStore
import coop.rchain.rspace.syntax.rspaceSyntaxKeyValueStoreManager
import coop.rchain.rspace.{Checkpoint, Match, RSpace}
import coop.rchain.shared.Log
import coop.rchain.store.InMemoryStoreManager
import monix.eval.Task
import monix.execution.Scheduler.Implicits.global
import org.scalatest.prop.PropertyChecks
import org.scalatest.{AppendedClues, FlatSpec, Matchers}

import scala.concurrent.duration._
import coop.rchain.rholang.externalservices.ExternalServices
import coop.rchain.rholang.externalservices.{
  GrpcClientMock,
  GrpcClientService,
  OllamaService,
  OllamaServiceMock,
  OpenAIService,
  OpenAIServiceImpl,
  OpenAIServiceMock
}
import coop.rchain.rholang.externalservices.TestExternalServices

class NonDeterministicProcessesSpec
    extends FlatSpec
    with Matchers
    with PropertyChecks
    with AppendedClues {

  // Enable OpenAI processes for testing
  System.setProperty("openai.enabled", "true")

  private def evaluateAndReplay(
      initialPhlo: Cost,
      term: String,
      externalServices: ExternalServices
  ): (EvaluateResult, EvaluateResult) = {

    implicit val logF: Log[Task]           = new Log.NOPLog[Task]
    implicit val metricsEff: Metrics[Task] = new metrics.Metrics.MetricsNOP[Task]
    implicit val noopSpan: Span[Task]      = NoopSpan[Task]()
    implicit val ms: Metrics.Source        = Metrics.BaseSource
    implicit val kvm                       = InMemoryStoreManager[Task]

    val evaluaResult = for {
      costLog <- costLog[Task]()
      cost    <- CostAccounting.emptyCost[Task](implicitly, metricsEff, costLog, ms)
      store   <- kvm.rSpaceStores
      spaces <- Resources.createRuntimes[Task](
                 store,
                 externalServices = externalServices
               )
      (runtime, replayRuntime, _) = spaces
      result <- {
        implicit def rand: Blake2b512Random = Blake2b512Random(Array.empty[Byte])
        runtime.evaluate(
          term,
          initialPhlo,
          Map.empty
        )(rand) >>= { playResult =>
          runtime.createCheckpoint >>= {
            case Checkpoint(root, log) =>
              replayRuntime.reset(root) >> replayRuntime.rig(log) >>
                replayRuntime.evaluate(
                  term,
                  initialPhlo,
                  Map.empty
                )(rand) >>= { replayResult =>
                replayRuntime.checkReplayData.as((playResult, replayResult))
              }
          }
        }
      }
    } yield result

    evaluaResult.runSyncUnsafe(7500.seconds)
  }

  private def assertReplayConsistency(
      initialPhlo: Cost,
      contract: String,
      testName: String,
      openAIService: OpenAIService = OpenAIServiceImpl.noOpInstance,
      grpcClient: GrpcClientService = GrpcClientService.noOpInstance,
      ollamaService: OllamaService = OllamaServiceMock.disabledService,
      printCosts: Boolean = false,
      expectError: Boolean = false
  ): Unit = {
    val (playResult, replayResult) =
      evaluateAndReplay(
        initialPhlo,
        contract,
        TestExternalServices(openAIService, grpcClient, ollamaService)
      )

    if (printCosts) {
      println(s"$testName - Initial Phlo: ${initialPhlo.value}")
      println(s"$testName - Play Cost: ${playResult.cost.value}")
      println(s"$testName - Replay Cost: ${replayResult.cost.value}")
      println(s"$testName - Cost Difference: ${playResult.cost.value - replayResult.cost.value}")
    }

    withClue(s"$testName - Play result should ${if (expectError) "have" else "not have"} errors: ") {
      if (expectError) {
        playResult.errors should not be empty
      } else {
        playResult.errors shouldBe empty
      }
    }

    withClue(
      s"$testName - Replay result should ${if (expectError) "have" else "not have"} errors: "
    ) {
      if (expectError) {
        replayResult.errors should not be empty
      } else {
        replayResult.errors shouldBe empty
      }
    }

    withClue(s"$testName - Replay cost should match play cost: ") {
      replayResult.cost shouldEqual playResult.cost
    }
  }

  "Replay" should "produce consistent costs without calling OpenAI service on successful execution" in {

    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new output, gpt4(`rho:ai:gpt4`) in { gpt4!(\"abc\", *output) }",
      "GPT4 process basic test",
      openAIService = OpenAIServiceMock.createSingleCompletionMock("gpt4 completion")
    )
  }

  it should "handle OutOfPhlogistonsError during replay with consistent cost accounting" in {
    assertReplayConsistency(
      Cost(1000),
      "new output, gpt4(`rho:ai:gpt4`) in { gpt4!(\"abc\", *output) }",
      "GPT4 process basic test with OutOfPhlogistonsError",
      openAIService = OpenAIServiceMock.createSingleCompletionMock("a" * 1000000),
      expectError = true
    )
  }

  it should "handle OpenAI service errors during replay with consistent error propagation" in {
    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new output, gpt4(`rho:ai:gpt4`) in { gpt4!(\"abc\", *output) }",
      "GPT4 process basic test with any other error",
      openAIService = OpenAIServiceMock.createErrorOnFirstCallMock(),
      expectError = true
    )
  }

  it should "produce consistent costs for DALL-E 3 without calling service on replay" in {
    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new output, dalle3(`rho:ai:dalle3`) in { dalle3!(\"a cat painting\", *output) }",
      "DALL-E 3 process basic test",
      openAIService =
        OpenAIServiceMock.createSingleDalle3Mock("https://example.com/generated-image.png")
    )
  }

  it should "handle OutOfPhlogistonsError during DALL-E 3 replay with consistent cost accounting" in {
    assertReplayConsistency(
      Cost(1000),
      "new output, dalle3(`rho:ai:dalle3`) in { dalle3!(\"a cat painting\", *output) }",
      "DALL-E 3 process with OutOfPhlogistonsError",
      openAIService = OpenAIServiceMock.createSingleDalle3Mock("https://" + "a" * 1000000 + ".png"),
      expectError = true
    )
  }

  it should "handle DALL-E 3 service errors during replay with consistent error propagation" in {
    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new output, dalle3(`rho:ai:dalle3`) in { dalle3!(\"a cat painting\", *output) }",
      "DALL-E 3 process with service error",
      openAIService = OpenAIServiceMock.createErrorOnFirstCallMock(),
      expectError = true
    )
  }

  it should "produce consistent costs for text-to-audio without calling service on replay" in {
    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new output, tts(`rho:ai:textToAudio`) in { tts!(\"Hello world\", *output) }",
      "Text-to-audio process basic test",
      openAIService =
        OpenAIServiceMock.createSingleTtsAudioMock("fake audio bytes".getBytes("UTF-8"))
    )
  }

  it should "handle OutOfPhlogistonsError during text-to-audio replay with consistent cost accounting" in {
    assertReplayConsistency(
      Cost(1000),
      "new output, tts(`rho:ai:textToAudio`) in { tts!(\"Hello world\", *output) }",
      "Text-to-audio process with OutOfPhlogistonsError",
      openAIService = OpenAIServiceMock.createSingleTtsAudioMock(Array.fill(1000000)(0.toByte)),
      expectError = true
    )
  }

  it should "handle text-to-audio service errors during replay with consistent error propagation" in {
    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new output, tts(`rho:ai:textToAudio`) in { tts!(\"Hello world\", *output) }",
      "Text-to-audio process with service error",
      openAIService = OpenAIServiceMock.createErrorOnFirstCallMock(),
      expectError = true
    )
  }

  it should "produce consistent costs for grpcTell on successful execution without calling service on replay" in {

    val grpcClientMock = GrpcClientMock.createSingleGrpcClientMock("localhost", 8080)

    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new grpcTell(`rho:io:grpcTell`) in { grpcTell!(\"localhost\", 8080, \"payload\") }",
      "grpcTell process basic test",
      grpcClient = grpcClientMock
    )

    grpcClientMock.wasCalled shouldBe true
  }

  it should "handle grpcTell connection errors during replay with consistent error propagation" in {

    val grpcClientMock = GrpcClientMock.createSingleGrpcClientMock("abc", 8081)
    assertReplayConsistency(
      Cost(Int.MaxValue),
      "new grpcTell(`rho:io:grpcTell`) in { grpcTell!(\"localhost\", 8080, \"payload\") }",
      "grpcTell process with connection error",
      grpcClient = grpcClientMock,
      expectError = true
    )

    grpcClientMock.wasCalled shouldBe true
  }
}

File: rholang/tests/accounting/non_deterministic_processes_spec.rs
// Tests for replay consistency of non-deterministic processes (OpenAI, gRPC, Ollama).
// Ensures that replays produce consistent costs and error handling for non-deterministic operations.

use crypto::rust::hash::blake2b512_random::Blake2b512Random;
use models::rhoapi::{BindPattern, ListParWithRandom, Par, TaggedContinuation};
use rholang::rust::interpreter::rho_runtime::RhoRuntimeImpl;
use rholang::rust::interpreter::test_utils::resources::create_runtimes_with_services;
use rholang::rust::interpreter::{
    accounting::costs::Cost,
    chromadb_service::create_noop_chromadb_service,
    external_services::ExternalServices,
    grpc_client_service::{GrpcClientMockConfig, GrpcClientService},
    interpreter::EvaluateResult,
    openai_service::{create_mock_openai_service, create_noop_openai_service, OpenAIMockConfig},
    rho_runtime::RhoRuntime,
};
use rspace_plus_plus::rspace::history::history_repository::HistoryRepository;
use rspace_plus_plus::rspace::shared::{
    in_mem_store_manager::InMemoryStoreManager, key_value_store_manager::KeyValueStoreManager,
};

use std::collections::HashMap;
use std::sync::Arc;

/// Helper to create external services with mock OpenAI and optional mock gRPC
fn create_test_external_services(
    openai_mock: OpenAIMockConfig,
    grpc_mock: Option<GrpcClientMockConfig>,
) -> ExternalServices {
    ExternalServices {
        openai: create_mock_openai_service(openai_mock),
        ollama: rholang::rust::interpreter::ollama_service::create_disabled_ollama_service(),
        grpc_client: match grpc_mock {
            Some(config) => GrpcClientService::new_mock(config),
            None => GrpcClientService::new_noop(),
        },
        chroma: create_noop_chromadb_service(),
        openai_enabled: true,
        ollama_enabled: false,
        is_validator: true,
    }
}

/// Helper to create external services with only mock gRPC
#[allow(dead_code)]
fn create_test_external_services_grpc(grpc_mock: GrpcClientMockConfig) -> ExternalServices {
    ExternalServices {
        openai: create_noop_openai_service(),
        ollama: rholang::rust::interpreter::ollama_service::create_disabled_ollama_service(),
        grpc_client: GrpcClientService::new_mock(grpc_mock),
        chroma: create_noop_chromadb_service(),
        openai_enabled: false,
        ollama_enabled: false,
        is_validator: true,
    }
}

/// Evaluate a term (play only), returning the result
async fn evaluate_with_mock_service(
    term: &str,
    external_services: ExternalServices,
) -> EvaluateResult {
    let mut kvm = InMemoryStoreManager::new();
    let store = kvm.r_space_stores().await.unwrap();
    let (runtime, _, _): (
        RhoRuntimeImpl,
        RhoRuntimeImpl,
        Arc<
            Box<
                dyn HistoryRepository<Par, BindPattern, ListParWithRandom, TaggedContinuation>
                    + Send
                    + Sync
                    + 'static,
            >,
        >,
    ) = create_runtimes_with_services(store, false, &mut Vec::new(), external_services).await;

    let rand = Blake2b512Random::create_from_bytes(&[]);
    let initial_phlo = Cost::create(i64::MAX, "test".to_string());

    runtime
        .evaluate(term, initial_phlo, HashMap::new(), rand)
        .await
        .expect("Evaluation failed")
}

/// Evaluate a term then replay it, returning (play_result, replay_result).
/// Evaluate-and-replay pattern for non-deterministic process testing:
///   1. Play: evaluate on normal runtime
///   2. Checkpoint: capture root hash + event log
///   3. Rig: reset replay runtime to root, load event log
///   4. Replay: evaluate same term on replay runtime
///   5. Verify: check_replay_data ensures all events were consumed
async fn evaluate_and_replay(
    term: &str,
    initial_phlo: Cost,
    external_services: ExternalServices,
) -> (EvaluateResult, EvaluateResult) {
    let mut kvm = InMemoryStoreManager::new();
    let store = kvm.r_space_stores().await.unwrap();
    let (mut runtime, mut replay_runtime, _): (
        RhoRuntimeImpl,
        RhoRuntimeImpl,
        Arc<
            Box<
                dyn HistoryRepository<Par, BindPattern, ListParWithRandom, TaggedContinuation>
                    + Send
                    + Sync
                    + 'static,
            >,
        >,
    ) = create_runtimes_with_services(store, false, &mut Vec::new(), external_services).await;

    let rand = Blake2b512Random::create_from_bytes(&[]);

    // Play phase
    let play_result = runtime
        .evaluate(term, initial_phlo.clone(), HashMap::new(), rand.clone())
        .await
        .expect("Play evaluation failed");

    // Checkpoint: captures root hash and event log
    let checkpoint = runtime.create_checkpoint().await;

    // Rig replay runtime with the event log from play
    replay_runtime
        .reset(&checkpoint.root)
        .await
        .expect("Replay reset failed");
    replay_runtime
        .rig(checkpoint.log)
        .await
        .expect("Replay rig failed");

    // Replay phase: same term, same phlo, same rand
    let replay_result = replay_runtime
        .evaluate(term, initial_phlo, HashMap::new(), rand)
        .await
        .expect("Replay evaluation failed");

    // Verify all replay events were consumed
    replay_runtime
        .check_replay_data()
        .await
        .expect("Replay data check failed: unconsumed events remain");

    (play_result, replay_result)
}

/// Assert that play and replay produce consistent results
fn assert_replay_consistency(
    play_result: &EvaluateResult,
    replay_result: &EvaluateResult,
    test_name: &str,
    expect_error: bool,
) {
    if expect_error {
        assert!(
            !play_result.errors.is_empty(),
            "{test_name}: play result should have errors"
        );
        assert!(
            !replay_result.errors.is_empty(),
            "{test_name}: replay result should have errors"
        );
    } else {
        assert!(
            play_result.errors.is_empty(),
            "{test_name}: play result should not have errors: {:?}",
            play_result.errors
        );
        assert!(
            replay_result.errors.is_empty(),
            "{test_name}: replay result should not have errors: {:?}",
            replay_result.errors
        );
    }

    assert_eq!(
        play_result.cost.value, replay_result.cost.value,
        "{test_name}: replay cost ({}) should match play cost ({})",
        replay_result.cost.value, play_result.cost.value
    );
}

// =====================================================
// Basic System Process Tests (no replay)
// These verify the mock services work correctly
// =====================================================

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_gpt4_mock_service_works() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::single_completion("gpt4 completion"), None);

    let result = evaluate_with_mock_service(
        r#"new output, gpt4(`rho:ai:gpt4`) in { gpt4!("abc", *output) }"#,
        external_services,
    )
    .await;

    assert!(
        result.errors.is_empty(),
        "GPT4 with mock service should not have errors: {:?}",
        result.errors
    );
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_gpt4_mock_error_returns_error() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::error_on_first_call(), None);

    let result = evaluate_with_mock_service(
        r#"new output, gpt4(`rho:ai:gpt4`) in { gpt4!("abc", *output) }"#,
        external_services,
    )
    .await;

    // The error should be captured by the NonDeterministicProcessFailure mechanism
    assert!(
        !result.errors.is_empty(),
        "GPT4 with error mock should have errors"
    );
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_dalle3_mock_service_works() {
    let external_services = create_test_external_services(
        OpenAIMockConfig::single_dalle3("https://example.com/generated-image.png"),
        None,
    );

    let result = evaluate_with_mock_service(
        r#"new output, dalle3(`rho:ai:dalle3`) in { dalle3!("a cat painting", *output) }"#,
        external_services,
    )
    .await;

    assert!(
        result.errors.is_empty(),
        "DALL-E 3 with mock service should not have errors: {:?}",
        result.errors
    );
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_dalle3_mock_error_returns_error() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::error_on_first_call(), None);

    let result = evaluate_with_mock_service(
        r#"new output, dalle3(`rho:ai:dalle3`) in { dalle3!("a cat painting", *output) }"#,
        external_services,
    )
    .await;

    assert!(
        !result.errors.is_empty(),
        "DALL-E 3 with error mock should have errors"
    );
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_tts_mock_service_works() {
    let external_services = create_test_external_services(
        OpenAIMockConfig::single_tts_audio(b"fake audio bytes".to_vec()),
        None,
    );

    let result = evaluate_with_mock_service(
        r#"new output, tts(`rho:ai:textToAudio`) in { tts!("Hello world", *output) }"#,
        external_services,
    )
    .await;

    assert!(
        result.errors.is_empty(),
        "Text-to-audio with mock service should not have errors: {:?}",
        result.errors
    );
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_tts_mock_error_returns_error() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::error_on_first_call(), None);

    let result = evaluate_with_mock_service(
        r#"new output, tts(`rho:ai:textToAudio`) in { tts!("Hello world", *output) }"#,
        external_services,
    )
    .await;

    assert!(
        !result.errors.is_empty(),
        "Text-to-audio with error mock should have errors"
    );
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_grpc_tell_mock_service_works() {
    let grpc_mock = GrpcClientMockConfig::create("localhost", 8080);
    let external_services = create_test_external_services_grpc(grpc_mock.clone());

    // gRPC tell uses 3 arguments (host, port, payload) without an ack channel
    let result = evaluate_with_mock_service(
        r#"new grpcTell(`rho:io:grpcTell`) in { grpcTell!("localhost", 8080, "payload") }"#,
        external_services,
    )
    .await;

    assert!(
        result.errors.is_empty(),
        "gRPC tell with mock service should not have errors: {:?}",
        result.errors
    );

    // Verify the mock was called
    assert!(grpc_mock.was_called(), "gRPC mock should have been called");
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn test_grpc_tell_mock_error_returns_error() {
    // Mock expects different host/port to trigger an error
    let grpc_mock = GrpcClientMockConfig::create("different_host", 9999);
    let external_services = create_test_external_services_grpc(grpc_mock.clone());

    // gRPC tell uses 3 arguments (host, port, payload) without an ack channel
    let result = evaluate_with_mock_service(
        r#"new grpcTell(`rho:io:grpcTell`) in { grpcTell!("localhost", 8080, "payload") }"#,
        external_services,
    )
    .await;

    // The gRPC error should be propagated
    assert!(
        !result.errors.is_empty(),
        "gRPC tell with error mock should have errors"
    );

    // Verify the mock was called
    assert!(grpc_mock.was_called(), "gRPC mock should have been called");
}

// =====================================================
// Replay Consistency Tests
// These verify play and replay produce identical costs
// (NonDeterministicProcessesSpec)
// =====================================================

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_gpt4_produces_consistent_costs() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::single_completion("gpt4 completion"), None);

    let (play, replay) = evaluate_and_replay(
        r#"new output, gpt4(`rho:ai:gpt4`) in { gpt4!("abc", *output) }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "GPT4 replay", false);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_gpt4_out_of_phlogistons_consistent_cost() {
    let external_services = create_test_external_services(
        OpenAIMockConfig::single_completion(&"a".repeat(1_000_000)),
        None,
    );

    let (play, replay) = evaluate_and_replay(
        r#"new output, gpt4(`rho:ai:gpt4`) in { gpt4!("abc", *output) }"#,
        Cost::create(1000, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "GPT4 OutOfPhlogistons replay", true);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_gpt4_service_error_consistent() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::error_on_first_call(), None);

    let (play, replay) = evaluate_and_replay(
        r#"new output, gpt4(`rho:ai:gpt4`) in { gpt4!("abc", *output) }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "GPT4 service error replay", true);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_dalle3_produces_consistent_costs() {
    let external_services = create_test_external_services(
        OpenAIMockConfig::single_dalle3("https://example.com/generated-image.png"),
        None,
    );

    let (play, replay) = evaluate_and_replay(
        r#"new output, dalle3(`rho:ai:dalle3`) in { dalle3!("a cat painting", *output) }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "DALL-E 3 replay", false);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_dalle3_service_error_consistent() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::error_on_first_call(), None);

    let (play, replay) = evaluate_and_replay(
        r#"new output, dalle3(`rho:ai:dalle3`) in { dalle3!("a cat painting", *output) }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "DALL-E 3 service error replay", true);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_tts_produces_consistent_costs() {
    let external_services = create_test_external_services(
        OpenAIMockConfig::single_tts_audio(b"fake audio bytes".to_vec()),
        None,
    );

    let (play, replay) = evaluate_and_replay(
        r#"new output, tts(`rho:ai:textToAudio`) in { tts!("Hello world", *output) }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "Text-to-audio replay", false);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_tts_service_error_consistent() {
    let external_services =
        create_test_external_services(OpenAIMockConfig::error_on_first_call(), None);

    let (play, replay) = evaluate_and_replay(
        r#"new output, tts(`rho:ai:textToAudio`) in { tts!("Hello world", *output) }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "Text-to-audio service error replay", true);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_grpc_tell_produces_consistent_costs() {
    let grpc_mock = GrpcClientMockConfig::create("localhost", 8080);
    let external_services = create_test_external_services_grpc(grpc_mock);

    let (play, replay) = evaluate_and_replay(
        r#"new grpcTell(`rho:io:grpcTell`) in { grpcTell!("localhost", 8080, "payload") }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "gRPC tell replay", false);
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn replay_grpc_tell_error_consistent() {
    let grpc_mock = GrpcClientMockConfig::create("different_host", 9999);
    let external_services = create_test_external_services_grpc(grpc_mock);

    let (play, replay) = evaluate_and_replay(
        r#"new grpcTell(`rho:io:grpcTell`) in { grpcTell!("localhost", 8080, "payload") }"#,
        Cost::create(i64::MAX, "test".to_string()),
        external_services,
    )
    .await;

    assert_replay_consistency(&play, &replay, "gRPC tell error replay", true);
}
